Identify ISA-JSON compliant investigations

somathias opened the pull request:
Adds a column to the investigations table to check if investigations are ISA-JSON compliant. Derived studies and assays inherit the flag from their parent investigation if they have the correct sample types associated. Furthermore, makes a initial simple attempt to entangle functionality related to handling ISA-JSON compliant elements from the Single Page view. 

More specifically, if the ISA-JSON compliance setting is enabled in the admin panel the following functionality is available:
- the form for creating investigations includes a check box for marking the investigation as ISA-JSON compliant.
- the behavior of buttons for handling Investigations, Studies and Assays depends on whether they are ISA-JSON compliant or not. This functionality is hence no longer dependent on viewing the Study in Single page or default view. ISA-JSON compliant investigations e.g. always display the button "Design Study" and non-ISA-JSON compliant investigations can be managed in SP view in the default way.
- ISA-JSON compliant Studies and Assays display the "Study Design"/"Assay Design" tabs also in default view. ~However, the dynamic table can currently only be populated when in Single Page view. In default view it remains empty.~
- Projects, Investigations, Studies and Assays all have buttons to switch between Single Page and default view and back.
- Once a Study has been associated with an Investigation, the is-ISA-JSON-compliant flag cannot be changed. 
- Default Studies/Assays that are created through the Create menu can only be linked to non-ISA-JSON compliant Investigations/Studies.

This PR addresses #1570, fixes #1355 and relates to #1657.

kdp-cloud commented:
Single Pages controller also the `export_isa function` [here](https://github.com/seek4science/seek/blob/bc101acd945425a0d22b569ee88dbc10cd575bb0/app/controllers/single_pages_controller.rb#L52-L63). You should delete this.

kdp-cloud commented on config/routes.rb:
```diff
@@ -405,6 +405,7 @@
     resources :people, :programmes, :projects, :assays, :studies, :models, :sops, :workflows, :data_files, :publications, :documents, only: [:index]
     member do
       get :export_isatab_json
+      get :export_isa, action: :export_isa
```
You should delete the route in the single_pages resource since it's not used anymore:
https://github.com/seek4science/seek/blob/bf3200c8d4c8c9b921269b5b233a76299394dc79/config/routes.rb#L754

somathias replied:
Good catch - fixed in [60fcdde](https://github.com/seek4science/seek/pull/1698/commits/60fcddee7fdfdeac3ece58fc81c8df8b8798b260)

kdp-cloud commented on app/views/assays/show.html.erb:
```diff
@@ -124,7 +134,7 @@
     <%= render partial: 'general/items_related_to', object: @assay %>
   <% end %>
-	<% if Seek::Config.isa_json_compliance_enabled && displaying_single_page? %>
+	<% if Seek::Config.isa_json_compliance_enabled && @assay.is_ISA_JSON_compliant %>
```
I don't know if you were planning on fixing it at a later stage but I just want to mention here that this will not work without modifying the dynamic table logic.
It requires a selected 'Samples table', which represents a sample type, from the treeview.

somathias replied:
Yes, we are aware that this requires future modification of the dynamic table logic to work (and that it hence currently does not).

kdp-cloud reviewed (commented):
Would be a good idea to add some tests:

- [x] Unit test for studies testing the ISA-JSON compliance. 
- [ ] Unit test for assays testing the ISA-JSON compliance. 
- [ ] Functional test testing the button text when an investigation is considered ISA-compliant.
- [ ] Functional test testing the button text when a study is considered ISA-compliant.
- [ ] Functional test testing the button text when an assay is considered ISA-compliant.

stuzart reviewed (changes requested):
a minor comment, but important, but to conform with the ruby code style guide, method names, and also database attributes, shouldn't have capitals. So
     `is_ISA_JSON_compliant`
should be
     `is_isa_json_compliant`
mysql itself isn't case senstive for column names, but rails will automaticaly be using that for it's attribute name - so will require either a new migration to change the column name or can be rolled back (`rake db:rollback`) and then change the existing migration before running the migration again

where calling `is_isa_json_compliant`, because it's boolean it should end with a ?, i.e `is_isa_json_compliant?` - which really helps interpreting the code

somathias commented:
I have now renamed the database attribute by adding a new migration and added a ? to the boolean method name to adhere to the ruby code style convention. See commits [c7a030b](https://github.com/seek4science/seek/pull/1698/commits/c7a030b8a6c68679be8a42004c9e13b113fdddaa) and [2a9db2b](https://github.com/seek4science/seek/pull/1698/commits/2a9db2b909b24b1aa65a68bc5b72ad81682092eb).
